nengo/builder/connection.py

```diff
@@ -85,7 +85,7 @@ def get_targets(conn, eval_points, dtype=None):
                     f"Building {conn}: Connection function returned "
                     "None. Cannot solve for decoders."
                 )
-            if not isinstance(out, (int, float, np.ndarray)):
+            if not isinstance(out, (int, float, np.ndarray, np.number)):
```

Comment:
It seems strange to me to check for the things that it can't be. Why not just `isinstance(out, (tuple, list))`? I think those are what we're looking for. Is there anything else?

Also, why `hstack`? Why not just have an `asarray` call on the output no matter what it is?

Reply:
There are lots of iterables, checking for tuple and list isn't sufficient. Checking if it's an iterable isn't ideal because ndarrays are iterables.

I don't think it's as simple as just making it into an array, but it might be. You can try that if you want.

Reply:
But what other iterables are supported by `hstack` and won't return an array with `dtype=object`? Numpy functions don't allow generators. If you look at our implementation of `is_array_like`, we basically assume list/tuple/ndarray. If we want to generally support more, I think we'd need to modify that function and possibly make some changes throughout the codebase.

Reply:
Yeah, I mean checking only tuple and list would be simpler. I'll switch.

Reply:
I looked into just using `asarray`, and there's at least one test where it breaks. We currently allow users to do something like `function=lambda x: [x, -x]` which, since `x` is a vector (even if it's a single-element vector), returns a 2D array when passed through `asarray`. So `hstack` is what we want.

As for what we allow, in general I'm in favour of being a bit more limited and forcing users to be a bit more explicit. You're right that somebody can implement a custom `Sequence` class and `hstack` appears to work with that. There are some situations where we get what I believe is unintended/unexpected behaviour though. For example, `np.hstack({1:2, 3:4}) == [1, 3]`, but a user returning a dictionary was probably trying to do something else, and to silently do something different than expected seems like an easy avenue for bugs. Note that this has just come up with using `hstack`; before the recent changes, returning something like a `dict` would fail.

Reply:
I'm not too concerned with cases like the `dict` where it should be obvious to anyone that what they're doing shouldn't work. Whitelisting tuple and list is probably fine; like you said, there are other places where we assume some types and the model probably wouldn't get to this point anyway. In the future we'll add explicit types to NengoCore and these things will be more clear.